drivers/sensor/ina23x/ina237_emul.c

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright (c) 2023 North River Systems Ltd
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Emulator for the TI INA237 I2C power monitor
```

Comment:
sad to see emulators being used to mock. why can't we just use FFF?

Reply:
I had followed the pattern of some of the other sensors which did emulators.  Also, in this case part of the complexity is the calculation done at compile time using the DT configuration for the struct `ina237_config::cal` property  in [drivers/sensor/ina23x/ina237.c - INA237_DRIVER_INIT](https://github.com/zephyrproject-rtos/zephyr/pull/60717/commits/3e24cddf54de848d38554f692598a048fa67ff6a#diff-9b554dcbdeedb8966d7d2168b09914d60c3dbc77afbf0d0c936d73f5c3b5f4f5).

The FFF could potentially be used to replace the `ina23x_reg_read*()' and 'ina23x_reg_write()` functions if the rest of the functionality works.  Do you have a suggestion on best to replace those functions that are in `zephyr/drivers/sensor/ina23x/ina23x_common.c`?  I am used to other mocking frameworks and have limited experience with FFF.  If we can replace the `ina23x_common.c` functions, then that would allow eliminating the I2C message decode/encode portion of the `ina237_emul.c` file.  The register array would probably still be used since that is the easiest way to capture multiple writes done by `ina237_init()`.

Reply:
I think it's better to fix the issue for now and open another PR to discuss testing approach.

Reply:
I tried to use FFF to replace the `ina23x_reg_read*()` and `ina23x_reg_write()` functions and I wasn't able to solve the problem where the INA237 driver did reads and writes before the unit test started and I could set the .custom_fake to the mocked function.  Not sure if there is a way to load the driver dynamically and if that is a possibility for testing.

I did do another version where I did link-time replacement of the `ina23x_reg_read*()` and `ina23x_reg_write()` functions with mocks.  It avoids have to emulate individual I2C transactions and doesn't create a specific emulator, but there is still a generic I2C emulator created behind the scenes.  It isn't any better than the emulated version and has the disadvantage of testing less.  You can take a look here:  [Version using ina23x_reg_readXX() and ina23x_reg_write() mocks](https://github.com/zephyrproject-rtos/zephyr/commit/a0617c227bda35141abdbd96b023eb910a56a2ee)

Taking a look at other options, the `zephyr/tests/drivers/sensor/icm42688` test case is interesting.  It uses FFF, but just for the interrupt mock and still uses an emulator.

Here are all of the sensor tests that I am aware of and a summary of what they are doing:

```
tests/drivers/sensor
├── accel                      I2C and SPI emulators behind the scenes, but doesn't test much
├── akm09918c            Uses I2C emulator in ./drivers/sensor/akm09918c/akm09918c_emul.c
├── generic                  Uses DEVICE_DEFINE() to create a dummy sensor_driver_api driver
├── icm42688              Uses SPI emulator in ./drivers/sensor/icm42688/icm42688_emul.c
├── ina237                   The driver we are talking about
└── sbs_gauge             I2C driver (but don't see any reference to an emulator).  Focuses on making sure driver simply
                                       returns a value vs. actually doing full testing.
```

So at this point, what would you like to do.  Should we:
1. merge this patch set as is and open a discussion in regards to the emulator
2. just fix the bugs and remove the unit tests and open a discussion
3. some other alternative

My preference is to merge this as is given that it follows the existing patterns.  If there is an easier way to test it, I am all for that as well since it will be useful in the future, but after spending an entire day trying different options, I'm still convinced the first approach here is the best as the alternatives test less and are just as complicated.

Reply:
@gmarull - please take another look and let me know if you are happy with this as-is or would like to refactor it.

Reply:
I moved the emulator out of the drivers/sensor/ina23x and included it with the test case.  Hopefully the cluttering of the drivers folder with the emulators was your concern and you don't mind then in the emulator.

Changes will be included in the next push to the PR.